### PR TITLE
[Wallet]Perform entire CWallet::TopUpKeyPool in a transaction.

### PR DESCRIPTION
**Pull request by pstratem**
Previously after each call to GenerateNewKey the log would be flushed causing a massive performance penalty.

### CODE REVIEW

**Comment by pstratem** [comment]
Massive performance improvement.

Adding 10k keys went from 10 minutes to <10 seconds.

**Comment by jonasschnelli** [comment]
Nice to see @pstratem work on the wallet!

Code-Review utACK.

**Comment by laanwj** [comment]
That's a very nice speed improvement.
Concept ACK.

**Comment by instagibbs** [comment]
My laptop goes from simply locking up a few minutes and crashing(something IO related that I can't figure out) to ~2 seconds.  

lightly tested ACK

**Comment by sipa** [comment]
Code review + concept ACK.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -121,9 +121,15 @@ bool CWallet::AddKeyPubKey(const CKey& secret, const CPubKey &pubkey)
     if (!fFileBacked)
         return true;
     if (!IsCrypted()) {
-        return CWalletDB(strWalletFile).WriteKey(pubkey,
-                                                 secret.GetPrivKey(),
-                                                 mapKeyMetadata[pubkey.GetID()]);
+        if (pwalletdb) {
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: I don't like the code duplication here at every call site, as well as the manual deleting setting-to-NULL of pwalletdb. What about a inner class for context management like
```
/** Database reference. Can be nested, in which case the outer reference will own the handle.
  */
class DBRef
{
private:
    bool owning;
public:
    DBRef(CWallet * wallet): owning(false)
    {
        if (!wallet->pwalletdb) {
            wallet->pwalletdb = new CWalletDB(wallet->strWalletFile);
            owning = true;
       }
    }
    CWalletDB &get() { return *wallet->pwalletdb;  }
    ~DBRef()
    {
         if(owning) {
             delete wallet->pwalletdb;
             wallet->pwalletdb = NULL;
         }
    }
};
```
Then you can change this to
```
DBRef ref;
ref.get().WriteKey(pubkey, secret.GetPrivKey(), mapKeyMetadata[pubkey.GetID());
```

Comment (pstratem) [inline src/wallet/wallet.cpp]: @laanwj yes i agree the if/else everywhere would get ugly very fast

**Comment by MarcoFalke** [comment]
Why is this closed?

**Comment by jgarzik** [comment]
Certainly would like to see some version of this go in...

**Pull request closed without merging**